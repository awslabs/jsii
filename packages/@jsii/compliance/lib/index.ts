export * from './parse-test-suite';
export * from './test-suite';
