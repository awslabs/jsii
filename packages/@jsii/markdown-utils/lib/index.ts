export * from './block';
export * from './element';
export * from './inline';
export * from './renderer';
export * from './renderers';
