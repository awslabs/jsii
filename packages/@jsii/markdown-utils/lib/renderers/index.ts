export * from './console-renderer';
export * from './json-renderer';
