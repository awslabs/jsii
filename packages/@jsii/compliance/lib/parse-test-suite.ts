import { BlockElement, CodeBlock, Document, Element, Heading, Renderer, Paragraph } from '@jsii/markdown-utils';
import { TestSuite, KernelTraceEntry, MessageDirection } from './test-suite';

/**
 * Parses a test suite from it's standard Markdown representation.
 *
 * @param source the source of the Markdown document
 *
 * @returns the parsed test suite.
 */
export function parseTestSuite(source: string): TestSuite | undefined {
  const document = Document.parse(source);
  return document.render(new TestSuiteRenderer());
}

class TestSuiteRenderer implements Renderer<TestSuite | undefined> {
  private testSuite?: TestSuite = undefined;
  private inTestSuite = false;
  private category?: string;
  private testName?: string;

  public renderHeading(heading: Heading) {
    if (heading.depth < 2) {
      this.inTestSuite = false;
      delete this.category;
      delete this.testName;
      return this.testSuite;
    } else if (heading.depth === 2) {
      this.inTestSuite = heading.text.trim() === 'Test Suite';
    }
    if (!this.inTestSuite || heading.depth <= 2) {
      return this.testSuite;
    }
    if (this.testSuite == null) {
      this.testSuite = {};
      delete this.category;
      delete this.testName;
    }
    if (heading.depth === 3) {
      this.category = heading.text.trim();
      this.testSuite[this.category] = this.testSuite[this.category] ?? {};
      delete this.testName;
    } else if (heading.depth === 4) {
      if (!this.category) {
        throw new Error(`Encountered a test case outside of a category: ${heading.text}`);
      }
      this.testName = heading.text;
      this.testSuite[this.category].testCases = this.testSuite[this.category].testCases ?? {};
      this.testSuite[this.category].testCases[this.testName] = this.testSuite[this.category].testCases[this.testName] ?? {};
    } else {
      throw new Error(`H${heading.depth} heading is deeper than the supported maximum of 4`);
    }
    return this.testSuite;
  }

  public renderCodeBlock(codeBlock: CodeBlock) {
    if (this.testSuite && this.category && this.testName) {
      const testCase = this.testSuite[this.category].testCases[this.testName];
      if (codeBlock.language !== '') {
        testCase.canonicalForm = codeBlock.text.trim();
      } else {
        testCase.kernelTrace = parseKernelTrace(codeBlock.text.trim());
      }
    }
    return this.testSuite;

    function parseKernelTrace(text: string) {
      const trace = new Array<KernelTraceEntry>();
      for (const line of text.split('\n')) {
        if (line.trim() === '' || /^\s*#/.test(line)) { continue; }
        const parts = /^([><])\s*(.+)$/.exec(line);
        if (!parts) {
          throw new Error(`Illegal kernel trace entry: ${line}`);
        }
        const [, directionChar, messageJson] = parts;
        trace.push({
          direction: directionChar === '>' ? MessageDirection.KernelToHost : MessageDirection.HostToKernel,
          message: JSON.parse(messageJson),
        });
      }
      return trace;
    }
  }

  public renderParagraph(paragraph: Paragraph) {
    if (this.testSuite && this.category && this.testName) {
      const testCase = this.testSuite[this.category].testCases[this.testName];
      if (!testCase.description) {
        testCase.description = paragraph.text.trim();
      } else {
        testCase.description += `\n\n${paragraph.text.trim()}`;
      }
    }
    return this.testSuite;
  }

  /* eslint-disable @typescript-eslint/member-ordering */
  public readonly renderBlockQuote = this.ignore;
  public readonly renderCode = this.ignore;
  public readonly renderHtmlBlock = this.ignore;
  public readonly renderHtmlInline = this.ignore;
  public readonly renderLineBreak = this.ignore;
  public readonly renderSoftBreak = this.ignore;
  public readonly renderText = this.ignore;
  public readonly renderThematicBreak = this.ignore;
  public readonly renderCustomBlock = this.ignore;
  public readonly renderCustomInline = this.ignore;
  public readonly renderDocument = this.ignore;
  public readonly renderEmph = this.ignore;
  public readonly renderImage = this.ignore;
  public readonly renderItem = this.ignore;
  public readonly renderLink = this.ignore;
  public readonly renderList = this.ignore;
  public readonly renderStrong = this.ignore;
  /* eslint-enable @typescript-eslint/member-ordering */

  private ignore(element: Element) {
    if (element instanceof BlockElement) {
      element.children.forEach(elt => elt.render(this));
    }
    return this.testSuite;
  }
}
