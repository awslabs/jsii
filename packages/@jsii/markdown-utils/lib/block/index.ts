import { Node } from 'commonmark';
import { BlockElement, Element } from '../element';

export * from './block-quote';
export * from './custom-block';
export * from './custom-inline';
export * from './document';
export * from './emph';
export * from './heading';
export * from './image';
export * from './item';
export * from './link';
export * from './list';
export * from './paragraph';
export * from './strong';

type BlockConstructor<T extends BlockElement> = new (node: Node, children: readonly Element[]) => T;

const constructors: Record<string, BlockConstructor<BlockElement>> = {};
for (const type of Object.values(module.exports)) {
  if (typeof type !== 'function') {
    // Ignore non-constructor exports
    continue;
  }
  if (typeof (type as any).commonMarkType === 'string') {
    const typeName = (type as any).commonMarkType;
    if (typeName in constructors) {
      throw new Error(`Duplicate definition for CommonMark type ${typeName}: ${constructors[typeName]} and ${type}`);
    }
    constructors[typeName] = (type as any);
  }
}

/** @internal */
export const blocks: { readonly [commonMarkType: string]: BlockConstructor<BlockElement> } = constructors;
