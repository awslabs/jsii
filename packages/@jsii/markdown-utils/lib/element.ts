import { Node } from 'commonmark';
import { Renderer } from './renderer';

/**
 * A Markdown document element.
 */
export abstract class Element {
  /**
   * Creates a new element.
   *
   * @param node the CommonMark node corresponding to this element.
   *
   * @internal
   */
  public constructor(protected readonly node: Node) {
  }

  /**
   * Retrieves the (un-decorated) text for this element.
   */
  public abstract get text(): string;

  /**
   * Renders this element.
   *
   * @param renderer the renderer to use.
   *
   * @returns the rendered result.
   */
  public render<T>(renderer: Renderer<T>): T {
    const handler = `render${this.constructor.name}`;
    if (handler in renderer) {
      return ((renderer as any)[handler])(this);
    }
    throw new Error(`Not implemented: ${renderer.constructor.name}#${handler}`);
  }
}

/**
 * A Markdown document block element.
 */
export abstract class BlockElement extends Element {
  /**
   * Creates a new block element.
   *
   * @param node the CommonMark node corresponding to this block element.
   * @param children the children of this block element.
   *
   * @internal
   */
  public constructor(node: Node, public readonly children: readonly Element[]) {
    super(node);
  }

  public get text() {
    return this.children.map(elt => elt.text).join('');
  }
}
