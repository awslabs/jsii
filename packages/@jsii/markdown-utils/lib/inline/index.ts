import { Node } from 'commonmark';
import { Element } from '../element';

export * from './code-block';
export * from './code';
export * from './html-block';
export * from './html-inline';
export * from './linebreak';
export * from './softbreak';
export * from './text';
export * from './thematic-break';

type Constructor<T extends Element> = new (node: Node) => T;

const constructors: Record<string, Constructor<Element>> = {};
for (const type of Object.values(module.exports)) {
  if (typeof type !== 'function') {
    // Ignore non-constructor exports
    continue;
  }
  if (typeof (type as any).commonMarkType === 'string') {
    const typeName = (type as any).commonMarkType;
    if (typeName in constructors) {
      throw new Error(`Duplicate definition for CommonMark type ${typeName}: ${constructors[typeName]} and ${type}`);
    }
    constructors[typeName] = (type as any);
  }
}

/** @internal */
export const inlines: { readonly [commonMarkType: string]: Constructor<Element> } = constructors;
