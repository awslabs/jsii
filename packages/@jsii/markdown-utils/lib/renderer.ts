import { BlockQuote, CustomBlock, CustomInline, Document, Emph, Heading, Image, Item, Link, List, Paragraph, Strong } from './block';
import { CodeBlock, Code, HtmlBlock, HtmlInline, LineBreak, SoftBreak, Text, ThematicBreak } from './inline';

export interface Renderer<T> {
  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderCodeBlock(element: CodeBlock): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderCode(element: Code): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderHtmlBlock(element: HtmlBlock): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderHtmlInline(element: HtmlInline): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderLineBreak(element: LineBreak): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderSoftBreak(element: SoftBreak): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderText(element: Text): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderThematicBreak(element: ThematicBreak): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderBlockQuote(element: BlockQuote): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderCustomBlock(element: CustomBlock): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderCustomInline(element: CustomInline): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderDocument(element: Document): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderEmph(element: Emph): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderHeading(element: Heading): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderImage(element: Image): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderItem(element: Item): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderLink(element: Link): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderList(element: List): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderParagraph(element: Paragraph): T;

  /**
   * Renders a Markdown element.
   *
   * @param element the element to be rendered.
   *
   * @returns the rendered result.
   */
  renderStrong(element: Strong): T;
}

