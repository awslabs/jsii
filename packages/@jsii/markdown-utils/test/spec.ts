import { readFileSync } from 'fs';
import { resolve } from 'path';
import { Document, JsonRenderer } from '../lib';

const asJson = new JsonRenderer();

describe(Document, () => {
  describe(Document.parse, () => {
    test('with an empty document', () => {
      const parsed = Document.parse('');
      expect(parsed.render(asJson)).toMatchSnapshot();
    });

    test('with nested lists', () => {
      const parsed = Document.parse(readSource('nested-list.md'));
      expect(parsed.render(asJson)).toMatchSnapshot();
    });

    test('with everything', () => {
      const parsed = Document.parse(readSource('everything.md'));
      expect(parsed.render(asJson)).toMatchSnapshot();
    });
  });
});

function readSource(name: string): string {
  return readFileSync(resolve(__dirname, 'md', name), { encoding: 'utf-8' });
}
