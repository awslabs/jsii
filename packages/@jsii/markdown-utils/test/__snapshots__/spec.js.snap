// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document parse with an empty document 1`] = `
Object {
  "document": Object {
    "children": Array [],
  },
}
`;

exports[`Document parse with everything 1`] = `
Object {
  "document": Object {
    "children": Array [
      Object {
        "heading": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "H1",
              },
            },
          ],
          "depth": 1,
        },
      },
      Object {
        "heading": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "H2",
              },
            },
          ],
          "depth": 2,
        },
      },
      Object {
        "heading": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "H3",
              },
            },
          ],
          "depth": 3,
        },
      },
      Object {
        "heading": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "H4",
              },
            },
          ],
          "depth": 4,
        },
      },
      Object {
        "heading": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "H5",
              },
            },
          ],
          "depth": 5,
        },
      },
      Object {
        "heading": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "H6",
              },
            },
          ],
          "depth": 6,
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Alternatively, for H1 and H2, an underline-ish style:",
              },
            },
          ],
        },
      },
      Object {
        "heading": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Alt-H1",
              },
            },
          ],
          "depth": 1,
        },
      },
      Object {
        "heading": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Alt-H2",
              },
            },
          ],
          "depth": 2,
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Emphasis, aka italics, with ",
              },
            },
            Object {
              "emph": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "asterisks",
                    },
                  },
                ],
              },
            },
            Object {
              "text": Object {
                "text": " or ",
              },
            },
            Object {
              "emph": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "underscores",
                    },
                  },
                ],
              },
            },
            Object {
              "text": Object {
                "text": ".",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Strong emphasis, aka bold, with ",
              },
            },
            Object {
              "strong": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "asterisks",
                    },
                  },
                ],
              },
            },
            Object {
              "text": Object {
                "text": " or ",
              },
            },
            Object {
              "strong": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "underscores",
                    },
                  },
                ],
              },
            },
            Object {
              "text": Object {
                "text": ".",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Combined emphasis with ",
              },
            },
            Object {
              "strong": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "asterisks and ",
                    },
                  },
                  Object {
                    "emph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "underscores",
                          },
                        },
                      ],
                    },
                  },
                ],
              },
            },
            Object {
              "text": Object {
                "text": ".",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Strikethrough uses two tildes. ~~Scratch this.~~",
              },
            },
          ],
        },
      },
      Object {
        "list": Object {
          "children": Array [
            Object {
              "item": Object {
                "children": Array [
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "First ordered list item",
                          },
                        },
                      ],
                    },
                  },
                ],
                "ordered": true,
              },
            },
            Object {
              "item": Object {
                "children": Array [
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "Another item",
                          },
                        },
                      ],
                    },
                  },
                ],
                "ordered": true,
              },
            },
          ],
        },
      },
      Object {
        "list": Object {
          "children": Array [
            Object {
              "item": Object {
                "children": Array [
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "Unordered sub-list.",
                          },
                        },
                      ],
                    },
                  },
                ],
                "ordered": false,
              },
            },
          ],
        },
      },
      Object {
        "list": Object {
          "children": Array [
            Object {
              "item": Object {
                "children": Array [
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "Actual numbers don",
                          },
                        },
                        Object {
                          "text": Object {
                            "text": "'",
                          },
                        },
                        Object {
                          "text": Object {
                            "text": "t matter, just that it",
                          },
                        },
                        Object {
                          "text": Object {
                            "text": "'",
                          },
                        },
                        Object {
                          "text": Object {
                            "text": "s a number",
                          },
                        },
                      ],
                    },
                  },
                ],
                "ordered": true,
              },
            },
            Object {
              "item": Object {
                "children": Array [
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "Ordered sub-list",
                          },
                        },
                      ],
                    },
                  },
                ],
                "ordered": true,
              },
            },
            Object {
              "item": Object {
                "children": Array [
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "And another item.",
                          },
                        },
                      ],
                    },
                  },
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "You can have properly indented paragraphs within list items. Notice the blank line above, and the leading spaces (at least one, but we",
                          },
                        },
                        Object {
                          "text": Object {
                            "text": "'",
                          },
                        },
                        Object {
                          "text": Object {
                            "text": "ll use three here to also align the raw Markdown).",
                          },
                        },
                      ],
                    },
                  },
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "To have a line break without a paragraph, you will need to use a trailing slash.",
                          },
                        },
                        Object {
                          "lineBreak": Object {},
                        },
                        Object {
                          "text": Object {
                            "text": "Note that this line is separate, but within the same paragraph.",
                          },
                        },
                        Object {
                          "lineBreak": Object {},
                        },
                        Object {
                          "text": Object {
                            "text": "(This is contrary to the typical GFM line break behaviour, where trailing spaces are not required.)",
                          },
                        },
                      ],
                    },
                  },
                ],
                "ordered": true,
              },
            },
          ],
        },
      },
      Object {
        "list": Object {
          "children": Array [
            Object {
              "item": Object {
                "children": Array [
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "Unordered list can use asterisks",
                          },
                        },
                      ],
                    },
                  },
                ],
                "ordered": false,
              },
            },
          ],
        },
      },
      Object {
        "list": Object {
          "children": Array [
            Object {
              "item": Object {
                "children": Array [
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "Or minuses",
                          },
                        },
                      ],
                    },
                  },
                ],
                "ordered": false,
              },
            },
          ],
        },
      },
      Object {
        "list": Object {
          "children": Array [
            Object {
              "item": Object {
                "children": Array [
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "Or pluses",
                          },
                        },
                      ],
                    },
                  },
                ],
                "ordered": false,
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "link": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "I",
                    },
                  },
                  Object {
                    "text": Object {
                      "text": "'",
                    },
                  },
                  Object {
                    "text": Object {
                      "text": "m an inline-style link",
                    },
                  },
                ],
                "destination": "https://www.google.com",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "link": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "I",
                    },
                  },
                  Object {
                    "text": Object {
                      "text": "'",
                    },
                  },
                  Object {
                    "text": Object {
                      "text": "m an inline-style link with title",
                    },
                  },
                ],
                "destination": "https://www.google.com",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "link": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "I",
                    },
                  },
                  Object {
                    "text": Object {
                      "text": "'",
                    },
                  },
                  Object {
                    "text": Object {
                      "text": "m a reference-style link",
                    },
                  },
                ],
                "destination": "https://www.mozilla.org",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "link": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "I",
                    },
                  },
                  Object {
                    "text": Object {
                      "text": "'",
                    },
                  },
                  Object {
                    "text": Object {
                      "text": "m a relative reference to a repository file",
                    },
                  },
                ],
                "destination": "../blob/master/LICENSE",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "link": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "You can use numbers for reference-style link definitions",
                    },
                  },
                ],
                "destination": "http://slashdot.org",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Or leave it empty and use the ",
              },
            },
            Object {
              "link": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "link text itself",
                    },
                  },
                ],
                "destination": "http://www.reddit.com",
              },
            },
            Object {
              "text": Object {
                "text": ".",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "URLs and URLs in angle brackets will automatically get turned into links.",
              },
            },
            Object {
              "softBreak": Object {},
            },
            Object {
              "text": Object {
                "text": "http://www.example.com or ",
              },
            },
            Object {
              "link": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "http://www.example.com",
                    },
                  },
                ],
                "destination": "http://www.example.com",
              },
            },
            Object {
              "text": Object {
                "text": " and sometimes",
              },
            },
            Object {
              "softBreak": Object {},
            },
            Object {
              "text": Object {
                "text": "example.com (but not on Github, for example).",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Some text to show that the reference links can follow later.",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Here",
              },
            },
            Object {
              "text": Object {
                "text": "'",
              },
            },
            Object {
              "text": Object {
                "text": "s our logo (hover to see the title text):",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Inline-style:",
              },
            },
            Object {
              "softBreak": Object {},
            },
            Object {
              "image": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "alt text",
                    },
                  },
                ],
                "url": "https://github.com/adam-p/markdown-here/raw/master/src/common/images/icon48.png",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Reference-style:",
              },
            },
            Object {
              "softBreak": Object {},
            },
            Object {
              "image": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "alt text",
                    },
                  },
                ],
                "url": "https://github.com/adam-p/markdown-here/raw/master/src/common/images/icon48.png",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Inline ",
              },
            },
            Object {
              "code": Object {
                "text": "code",
              },
            },
            Object {
              "text": Object {
                "text": " has ",
              },
            },
            Object {
              "code": Object {
                "text": "back-ticks around",
              },
            },
            Object {
              "text": Object {
                "text": " it.",
              },
            },
          ],
        },
      },
      Object {
        "codeBlock": Object {
          "language": "javascript",
          "text": [Function],
        },
      },
      Object {
        "codeBlock": Object {
          "language": "python",
          "text": [Function],
        },
      },
      Object {
        "codeBlock": Object {
          "language": "",
          "text": [Function],
        },
      },
      Object {
        "blockQuote": Object {
          "children": Array [
            Object {
              "parapgraph": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "Blockquotes are very handy in email to emulate reply text.",
                    },
                  },
                  Object {
                    "softBreak": Object {},
                  },
                  Object {
                    "text": Object {
                      "text": "This line is part of the same quote.",
                    },
                  },
                ],
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Quote break.",
              },
            },
          ],
        },
      },
      Object {
        "blockQuote": Object {
          "children": Array [
            Object {
              "parapgraph": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "This is a very long line that will still be quoted properly when it wraps. Oh boy let",
                    },
                  },
                  Object {
                    "text": Object {
                      "text": "'",
                    },
                  },
                  Object {
                    "text": Object {
                      "text": "s keep writing to make sure this is long enough to actually wrap for everyone. Oh, you can ",
                    },
                  },
                  Object {
                    "emph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "put",
                          },
                        },
                      ],
                    },
                  },
                  Object {
                    "text": Object {
                      "text": " ",
                    },
                  },
                  Object {
                    "strong": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "Markdown",
                          },
                        },
                      ],
                    },
                  },
                  Object {
                    "text": Object {
                      "text": " into a blockquote.",
                    },
                  },
                ],
              },
            },
          ],
        },
      },
      Object {
        "htmlBlock": Object {
          "text": "<dl>
  <dt>Definition list</dt>
  <dd>Is something people use sometimes.</dd>",
        },
      },
      Object {
        "htmlBlock": Object {
          "text": "  <dt>Markdown in HTML</dt>
  <dd>Does *not* work **very** well. Use HTML <em>tags</em>.</dd>
</dl>",
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Three or more...",
              },
            },
          ],
        },
      },
      Object {
        "thematicBreak": Object {},
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Hyphens",
              },
            },
          ],
        },
      },
      Object {
        "thematicBreak": Object {},
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Asterisks",
              },
            },
          ],
        },
      },
      Object {
        "thematicBreak": Object {},
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Underscores",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "Here",
              },
            },
            Object {
              "text": Object {
                "text": "'",
              },
            },
            Object {
              "text": Object {
                "text": "s a line for us to start with.",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "This line is separated from the one above by two newlines, so it will be a ",
              },
            },
            Object {
              "emph": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "separate paragraph",
                    },
                  },
                ],
              },
            },
            Object {
              "text": Object {
                "text": ".",
              },
            },
          ],
        },
      },
      Object {
        "parapgraph": Object {
          "children": Array [
            Object {
              "text": Object {
                "text": "This line is also a separate paragraph, but...",
              },
            },
            Object {
              "softBreak": Object {},
            },
            Object {
              "text": Object {
                "text": "This line is only separated by a single newline, so it",
              },
            },
            Object {
              "text": Object {
                "text": "'",
              },
            },
            Object {
              "text": Object {
                "text": "s a separate line in the ",
              },
            },
            Object {
              "emph": Object {
                "children": Array [
                  Object {
                    "text": Object {
                      "text": "same paragraph",
                    },
                  },
                ],
              },
            },
            Object {
              "text": Object {
                "text": ".",
              },
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`Document parse with nested lists 1`] = `
Object {
  "document": Object {
    "children": Array [
      Object {
        "list": Object {
          "children": Array [
            Object {
              "item": Object {
                "children": Array [
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "List item one.",
                          },
                        },
                      ],
                    },
                  },
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "List item one continued with a second paragraph followed by an",
                          },
                        },
                        Object {
                          "softBreak": Object {},
                        },
                        Object {
                          "text": Object {
                            "text": "Indented block.",
                          },
                        },
                      ],
                    },
                  },
                  Object {
                    "codeBlock": Object {
                      "language": null,
                      "text": [Function],
                    },
                  },
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "List item continued with a third paragraph.",
                          },
                        },
                      ],
                    },
                  },
                ],
                "ordered": true,
              },
            },
            Object {
              "item": Object {
                "children": Array [
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "List item two continued with an open block.",
                          },
                        },
                      ],
                    },
                  },
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "This paragraph is part of the preceding list item.",
                          },
                        },
                      ],
                    },
                  },
                  Object {
                    "list": Object {
                      "children": Array [
                        Object {
                          "item": Object {
                            "children": Array [
                              Object {
                                "parapgraph": Object {
                                  "children": Array [
                                    Object {
                                      "text": Object {
                                        "text": "This list is nested and does not require explicit item continuation.",
                                      },
                                    },
                                  ],
                                },
                              },
                              Object {
                                "parapgraph": Object {
                                  "children": Array [
                                    Object {
                                      "text": Object {
                                        "text": "This paragraph is part of the preceding list item.",
                                      },
                                    },
                                  ],
                                },
                              },
                            ],
                            "ordered": true,
                          },
                        },
                        Object {
                          "item": Object {
                            "children": Array [
                              Object {
                                "parapgraph": Object {
                                  "children": Array [
                                    Object {
                                      "text": Object {
                                        "text": "List item b.",
                                      },
                                    },
                                  ],
                                },
                              },
                            ],
                            "ordered": true,
                          },
                        },
                      ],
                    },
                  },
                  Object {
                    "parapgraph": Object {
                      "children": Array [
                        Object {
                          "text": Object {
                            "text": "This paragraph belongs to item two of the outer list.",
                          },
                        },
                      ],
                    },
                  },
                ],
                "ordered": true,
              },
            },
          ],
        },
      },
    ],
  },
}
`;
